File: medium/323-Number-of-Connected-Components-in-an-Undirected-Graph/UnionFind.java
class Solution {
    public int countComponents(int n, int[][] edges) {
        UnionFind uf = new UnionFind(n);
        for (int[] edge: edges) {
            uf.merge(edge[0], edge[1]);
        }
        
        int numComp = 0;
        for (int i = 0; i < n; i++) {
            if (uf.isRoot(i)) ++numComp;
        }
        return numComp;
    }
}

class UnionFind {
    private int[] parent;
    private int[] size;
    
    public UnionFind(int num) {
        parent = new int[num];        
        size = new int[num];
        
        for (int i = 0; i < num; i++) {
            parent[i] = i;
            size[i] = 1;
        }
    }
    
    public boolean isRoot(int i) {
        return i == parent[i];
    }
    
    public int root(int i) {
        if (!isRoot(i)) {
            parent[i] = root(parent[i]);
        }
        return parent[i];
    }
    
    public void merge(int i, int j) {
        i = root(i);
        j = root(j);
        
        if (i == j) return;
        
        if (size[i] < size[j]) {
            int tmp = j;
            j = i;
            i = tmp;
        }
        
        parent[j] = i;
        size[i] += size[j];
    }
}

File: medium/323-Number-of-Connected-Components-in-an-Undirected-Graph/UnionFind.kt
class UnionFind(num: Int) {
    val parent: IntArray
    val size: IntArray
    
    init {
        parent = IntArray(num) { it }
        size = IntArray(num) {1}
    }
    
    fun isRoot(i: Int): Boolean {
        return i == parent[i]
    }
    
    fun root(i: Int): Int {
        if (!isRoot(i)) {
            parent[i] = root(parent[i])
        }
        return parent[i]
    }
    
    fun merge(i: Int, j: Int) {
        var iRoot = root(i)
        var jRoot = root(j)
        
        if (iRoot == jRoot) {
            return
        }
        
        if (size[iRoot] < size[jRoot]) {
            val tmp = jRoot
            jRoot = iRoot
            iRoot = tmp
        }
        
        parent[jRoot] = iRoot
        size[iRoot] += size[jRoot]
    }
}

class Solution {
    fun countComponents(n: Int, edges: Array<IntArray>): Int {
        val unionFind = UnionFind(n)
        
        for (edge in edges) {
            unionFind.merge(edge[0], edge[1])
        }
        
        return (0 until n).filter { unionFind.isRoot(it) }.count()
        
    }
}
